iwyu.cc

```diff
@@ -2043,21 +2043,87 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     if (CanIgnoreCurrentASTNode())
       return true;
 
-    // We assume that all type traits with >= 2 arguments, except '__is_same',
-    // require full type information even for pointer types. For example,
-    // this is the case for `__is_convertible_to` trait.
-    if (expr == nullptr || expr->getNumArgs() < 2 ||
-        expr->getTrait() == TypeTrait::BTT_IsSame) {
+    if (expr == nullptr || expr->getNumArgs() < 2)
       return true;
-    }
 
-    for (const TypeSourceInfo* arg : expr->getArgs()) {
-      QualType qual_type = arg->getType();
-      const Type* type = qual_type.getTypePtr();
-      ReportTypeUse(CurrentLoc(), type, DerefKind::RemoveRefsAndPtr);
-    }
+    const Type* lhs_type = expr->getArg(0)->getType().getTypePtr();
+    const Type* rhs_type = nullptr;
+    if (expr->getNumArgs() == 2)
+      rhs_type = expr->getArg(1)->getType().getTypePtr();
+
+    // Most of the binary type traits require the types to be complete types or
+    // arrays of unknown bound for compilable code. It means that compilers
+    // accept pointers or references to fwd-declared types. However, in certain
+    // cases it may cause a UB (C++20 [meta.rqmts]p5) because the full
+    // pointed-to type is needed to determine the trait value correctly, so it's
+    // better for IWYU to require the full type info in such situations. It is
+    // important not to suggest a full type when it is redundant for computing
```

Comment:
I don't understand the last sentence, particularly "redundant for computing the correct value"?

Reply:
I mean: in cases like `std::is_assignable_v<Token*&, Token*>`, the result is always `true` regardless of what `Token` actually is (like in #1615). No full type info is needed.

Reply:
Sorry, I forgot. We decided to use the word "redundant" when an `#include` is already somewhere in the tree, didn't we? If so, "unnecessary" fits better here, indeed.